### PR TITLE
feat(tree): add data sources for tree

### PR DESCRIPTION
**Pull request by tinayuangao**

### CODE REVIEW

File: src/lib/tree/data-source/flat-data-source.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+ *    expandable: false,
+ *    level: 2
+ *  },
+ *  NodeTwo: {
+  *   key: 'Pear',
+  *   expandable: false,
+  *   level: 2
+  * }
+ */
+export class MatTreeFlattener<T, F> {
```
Comment (jelbourn) [inline src/lib/tree/data-source/flat-data-source.ts]: I don't quite follow what `T` and `F` are. Is one the type of node and one the type of data? If so, should the type of node include an `extends`?

File: src/lib/tree/data-source/nested-data-source.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+/**
+ * Data source for nested tree.
+ *
+ * The data source for nested tree doesn't have to consider node flattener, or the way to expand
+ * or collapse. The expansion/collapsion will be handled by TreeControl and each non-leaf node.
+ */
+export class MatTreeNestedDataSource<T> implements DataSource<T> {
+  _renderedData = new BehaviorSubject<T[]>([]);
+  get renderedData(): T[] { return this._renderedData.value; }
```
Comment (jelbourn) [inline src/lib/tree/data-source/nested-data-source.ts]: I don't quite follow what `renderedData` is needed/used for

File: src/lib/tree/data-source/flat-data-source.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+/**
+ * Data source for flat tree.
+ * The data source need to handle expansion/collapsion of the tree node and change the data feed
+ * to `MatTree`.
+ * The nested tree nodes of type `T` are flattened through `MatTreeFlattener`, and converted
+ * to type `F` for `MatTree` to consume.
+ */
+export class MatTreeFlatDataSource<T, F> implements DataSource<F> {
+  _flattenedData = new BehaviorSubject<F[]>([]);
+  get flattenedData() { return this._flattenedData.value; }
```
Comment (jelbourn) [inline src/lib/tree/data-source/flat-data-source.ts]: Why expose `flattenedData` in addition to the data stream coming back from `connect`?

File: src/lib/tree/data-source/flat-data-source.ts (hunk 1 above)
Comment (tinayuangao) [reply src/lib/tree/data-source/flat-data-source.ts]: Discussed offline and added comments about this.

File: src/lib/tree/data-source/nested-data-source.ts (hunk 2 above)
Comment (tinayuangao) [reply src/lib/tree/data-source/nested-data-source.ts]: Removed. `data` is enough

File: src/lib/tree/data-source/flat-data-source.ts (hunk 3 above)
Comment (tinayuangao) [reply src/lib/tree/data-source/flat-data-source.ts]: Removed getters

**Comment by jelbourn** [comment]
Poke @andrewseguin

**Pull request merged**